chipzoller/hugo-clarity
# Allow for subdirs in site address

This PR...

## Changes / fixes

- Fixes #322
- Fixes #325 

Changes:

- [Replaces `parentURL` and `rootURL` variables in JS with `baseURL`](https://github.com/chipzoller/hugo-clarity/pull/324/files#diff-e78f4acd103c72d6dcf89e1e42484ceca4d0fe11e725a7b5dff6d482af833b26) derived from Hugo config (in a config file or at the command line with `--baseURL`)
- [Changes the URL construction of the calls to `index.json` coming from search](https://github.com/chipzoller/hugo-clarity/pull/324/files#diff-215523d38d6c15f1b60a7afc252cc2841414757f34385b89900a7b13bb92d162). I'm not sure why this was failing when changing the `rootURL`/`baseURL` value before, but simply concatenating the values seems to work.
- Replaces a hardcoded call to `images` with the configurable image directory in the image-related [layout](https://github.com/chipzoller/hugo-clarity/pull/324/files#diff-ca8843b42c6efe94a57c66f9aa842be46241b02bb5db66cd2d6cb6ea5143a012) [partials](https://github.com/chipzoller/hugo-clarity/pull/324/files#diff-6df0b4e56673418819f3d04c957716668d3fc541d64d10bb50b3ad8c30381599).
- Updates the example site's [posts](https://github.com/chipzoller/hugo-clarity/pull/324/files#diff-fd6efad06fd380306decff299e3be5bd8a1144ac1c725e8bb84a5db4ebb7f1f3) to ensure images load even in a site within a subdir (and they still load fine in a site at a root domain).

Adds:

- [Provides configurable `imagesDir`, `fontsDir` and `iconsDir` variables](https://github.com/chipzoller/hugo-clarity/pull/324/files#diff-8b2e5d757c1d24d93c6691ffa8510b1c4f25f09e2770d53604e815d6182a619b) and normalizes the names.
- Passes those values, plus `baseURL`, to Sass as variables. Hugo's "[ExecuteAsTemplate](https://gohugo.io/hugo-pipes/resource-from-template/)" option only works on the primary Sass file, not partials (ref [here](https://blog.fullstackdigital.com/how-to-use-hugo-template-variables-in-scss-files-in-2018-b8a834accce) and [here](https://www.regisphilibert.com/blog/2018/07/hugo-pipes-and-asset-processing-pipeline/)), so [these are defined in `main.sass`](https://github.com/chipzoller/hugo-clarity/pull/324/files#diff-299d4db85d45383ffae69b01dee0d8d35618252f8e4ca40d084730bc780b0a28). This means these variables are also available to end-users in their `_override.s{a|c}ss` and `_custom.s{a|c}ss` files!
- Uses the Sass variables in the partials to define paths to fonts and images, including the subheader-permalink icon, which is moved from JS to Sass to make it easier to override for end-users. (Otherwise they'd have to override the entire JS function just to change the image.)

## Testing this PR

1. Try using the theme with a site address containing a subdirectory (e.g. `hugo server --baseURL http://localhost:1313/foo/`)
2. Try using the theme with a site address at a root domain (e.g. `hugo server`)

In each, check:

- The ability to perform searches
- The presence of the light/dark mode switch icons
- The presence of the correct custom fonts
- On the "Markdown Syntax Guide" post, ensure that
  - all images load
  - the permalink icon (chain link) appears when hovering over subheadings
  - the icons appear in the code block when hovering (copy, number, etc.)
- Optionally open a web debug console and see if any file references are returning 404s

## Checklist

_Ensure you have checked off the following before submitting your PR._

- [x] tested locally with the [latest release of Hugo](https://github.com/gohugoio/hugo/releases). This requirement is [a standard](https://github.com/gohugoio/hugoThemes#theme-maintenance)
- [x] added new dependencies (n/a)
